crates/apollo-parser/src/error.rs

```diff
@@ -34,10 +34,37 @@ use std::fmt;
 /// [annotate_snippets]: https://github.com/apollographql/apollo-rs/blob/a7f616454a53dcb8496725ceac6c63eacddefb2c/crates/apollo-parser/examples/annotate_snippet.rs
 /// [miette crate]: https://docs.rs/miette/3.2.0/miette/index.html
 
-#[derive(PartialEq, Eq, Clone)]
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub(crate) enum ErrorData {
+    Eof,
+    LimitExceeded,
+    Text(String),
+}
+
+impl ErrorData {
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Eof | Self::LimitExceeded => 0,
+            Self::Text(text) => text.len(),
+        }
+    }
+}
+
+impl fmt::Display for ErrorData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Eof => write!(f, "EOF"),
+            Self::LimitExceeded => Ok(()),
+            Self::Text(text) => write!(f, "{}", text),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, thiserror::Error)]
+#[error("ERROR@{index}:{} {message:?} {data}", .index + .data.len())]
```

Comment:
this reminds me, we should implement a better `Display` for just the parser errors. We didn't implement it earlier while still deciding on how diagnostics will look like in the compiler, but now those are pretty certain, so these can have a display too.